File: luogu/糖果.py
import sys, heapq
from collections import *
from functools import lru_cache
sys.setrecursionlimit(10**6)



def main():

    def dfs(cur,parent,d):
        if cur in people:
            return d, cur
        res = [(float('inf'),float('inf'))]
        for nxt in g[cur]:
            if nxt == parent: continue
            nxtd, nxtcur = dfs(nxt,cur,d+1)
            res.append((nxtd,nxtcur))
        return min(res)



    # sys.stdin = open('input.txt', 'r')
    # print(input().split(' '))
    n, m = map(int, input().split(' '))
    g = defaultdict(list)
    for _ in range(n-1):
        a, b = map(int, input().split(' '))
        g[a].append(b)
        g[b].append(a)

    people = set(list(map(int,input().split(' '))))
    # print(g)
    # print(people)
    print(dfs(1,0,0)[-1])






if __name__ == "__main__":
    main()

File: luogu/糖果.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <algorithm>
#include <vector>
using namespace std;

const int MAXN = 1e6 + 5;
const int INF = 0x7fffffff;

vector<int> g[MAXN];
int n,m,a,b,people[MAXN],power[MAXN],depth[MAXN];

void dfs(int cur, int parent, int d){
    if(people[cur]==1){
        power[cur]=cur;
        depth[cur]=d;
        return;
    }
    power[cur]=INF;depth[cur]=INF;
    for(auto nxt: g[cur]){
        if(nxt==parent)continue;
        dfs(nxt, cur, d+1);
        if ((depth[nxt]<depth[cur])||(depth[nxt]==depth[cur] && power[nxt]<power[cur])) {
            depth[cur]=depth[nxt];
            power[cur]=power[nxt];
        }
    }
    return;
}


signed main(void)
{
//    freopen("input.txt", "r", stdin);
    memset(people, 0, sizeof(people));
    memset(power, 0, sizeof(power));
    memset(depth, 0, sizeof(depth));
    scanf("%d %d", &n, &m);
    for (int i=1; i<=n-1; i++) {
        scanf("%d %d",&a,&b);
        g[a].push_back(b);
        g[b].push_back(a);
    }
    for (int i=0; i<m; i++) {
        int p;
        scanf("%d",&p);
        people[p]=1;
    }
    dfs(1,0,0);
    printf("%d\n",power[1]);
    
    return 0;
}
